pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -67,17 +67,67 @@ type topologyPairsMaps struct {
 	podToTopologyPairs map[string]topologyPairSet
 }
 
-// topologyPairsPodSpreadMap combines topologyKeyToMinPodsMap and topologyPairsMaps
+type criticalPath struct {
+	// topologyValue denotes the topology value mapping to topology key.
+	topologyValue string
+	// matchNum denotes the number of matching pods.
+	matchNum int32
+}
+
+// CAVEAT: the reason that `[2]criticalPath` can work is based on the implementation of current
+// preemption algorithm - we only preempt pods on the same node, instead of pods on multiple nodes.
+// If we plan to turn to a more complex algorithm like "arbitrary pods on multiple nodes", this
+// structure needs to be revisited.
+type criticalPaths [2]criticalPath
+
+func newCriticalPaths() *criticalPaths {
+	return &criticalPaths{{matchNum: math.MaxInt32}, {matchNum: math.MaxInt32}}
+}
+
+func (paths *criticalPaths) update(tpVal string, num int32) {
+	// first verify if `tpVal` exists or not
+	i := -1
+	if tpVal == paths[0].topologyValue {
+		i = 0
+	} else if tpVal == paths[1].topologyValue {
+		i = 1
+	}
+
+	if i >= 0 {
+		// `tpVal` exists
+		paths[i].matchNum = num
+		j := (i + 1) % 2 // the other index
+		// adjust i and j to make sure paths[0] always holds the minimum match num.
+		if (i < j && paths[i].matchNum > paths[j].matchNum) ||
+			(i > j && paths[i].matchNum < paths[j].matchNum) {
+			// swap paths[0] and paths[1]
+			paths[0], paths[1] = paths[1], paths[0]
+		}
+	} else {
+		// `tpVal` doesn't exist
+		if num < paths[0].matchNum {
+			// update paths[1] with paths[1]
+			paths[1] = paths[0]
+			// update paths[0]
+			paths[0].topologyValue, paths[0].matchNum = tpVal, num
+		} else if num < paths[1].matchNum {
+			// update paths[1]
+			paths[1].topologyValue, paths[1].matchNum = tpVal, num
+		}
+	}
+}
+
+// podSpreadCache combines tpKeyToCriticalPaths and tpPairToMatchNum
 // to represent:
-// (1) minimum number of pods matched on the spread constraints.
-// (2) how existing pods match incoming pod on its spread constraints.
-type topologyPairsPodSpreadMap struct {
-	// This map is keyed with a topology key, and valued with minimum number
-	// of pods matched on that topology domain.
-	// TODO(Huang-Wei): refactor to {tpKey->tpValSet(or tpValSlice)}
-	topologyKeyToMinPodsMap map[string]int32
-	// TODO(Huang-Wei): refactor to {tpPair->count, podName->tpPairSet(optional)}
-	*topologyPairsMaps
+// (1) critical paths where the least pods are matched on each spread constraint.
+// (2) number of pods matched on each spread constraint.
+type podSpreadCache struct {
+	// We don't record all critical paths here.
+	// If there is only one criticalPath, criticalPaths[0] always holds the minimum matchNum;
+	// If there are multiple criticalPaths, keep arbitrary 2 of them holding the minimum matchNum.
```

Comment:
That sounds too short... How about:

```
	// We record 2 critical paths instead of all critical paths here.
	// criticalPaths[0].matchNum always holds the minimum matching number.
	// criticalPaths[1].matchNum is always greater or equal to criticalPaths[0].matchNum,
	// but it's not guaranteed to be the 2nd minimum match number.
```